File: 01.Pattern-Sliding-Window/06.no_repeat_substring/non_repeat_substring.js
function non_repeat_substring(str) {
  let windowStart = 0,
    maxLength = 0,
    charIndexMap = {};

  // try to extend the range [windowStart: windowEnd]
  for (let windowEnd = 0; windowEnd < str.length; windowEnd++) {
    const rightChar = str[windowEnd];
    // if the map already contains the 'rightChar', shrink the window from the beginning so that
    // we have only one occurrence of 'rightChar'
    if (rightChar in charIndexMap) {
      // this is tricky; in the current window, we will not have any 'rightChar' after its previous
      // and if 'windowStart' is already ahdead the last index of 'rightChar'
      windowStart = Math.max(windowStart, charIndexMap[rightChar] + 1);
    }
    // insert the 'rightChar' into the map
    charIndexMap[rightChar] = windowEnd;
    // remember the maximum length so far
    maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
  }
  return maxLength;
}


console.log(`Length of the longest subbstring: ${non_repeat_substring('aabccbb')}`)
console.log(`Length of the longest subbstring: ${non_repeat_substring('abbbb')}`)
console.log(`Length of the longest subbstring: ${non_repeat_substring('abccde')}`)

File: 01.Pattern-Sliding-Window/06.no_repeat_substring/non_repeat_substring.py
def non_repeat_substring(str):
    window_start = 0
    max_length = 0
    char_index_map = {}

    # try to extend the range [windowStart: windowEnd]
    for window_end in range(len(str)):
        right_char = str[window_end]
        # if the map already contains the 'right_char', shrink the window from the beginning so that
        # we have only one crrurence of 'right_char'
        if right_char in char_index_map:
            # this is tricky; in the current window, we will not have any 'right_char' after its previous
            # and if 'window_start' is already ahead of the last index of 'right_char', we'll keep 'window_end'
            window_start = max(window_start, char_index_map[right_char] + 1)

        # insert the 'right_char' into the map
        char_index_map[right_char] = window_end
        # remember the maximum length so far
        max_length = max(max_length, window_end - window_start + 1)

    return max_length


def main():
    print("Length of the longest substring: " +
          str(non_repeat_substring("aabccbbb")))
    print("Length of the longest substring: " +
          str(non_repeat_substring("abbbb")))
    print("Length of the longest substring: " +
          str(non_repeat_substring("abccde")))


main()
